### PR TITLE
Lifted structure loss

### PR DESCRIPTION
**Pull request by Lorenzobattistela**
This PR refers to Issue #102 and it is implementing the LSL described in the article https://arxiv.org/pdf/1511.06452.pdf

### CODE REVIEW

**Comment by Lorenzobattistela** [comment]
@owenvallis fixed the CLA problem, same changes from before.

File: tensorflow_similarity/losses/lifted_structure_loss.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    It's values represent the classes associated with the examples as
+    **integer values**.
+    `y_pred` must be 2-D float `Tensor` of L2 normalized embedding vectors.
+    You can use the layer `tensorflow_similarity.layers.L2Embedding()` as the
+    last layer of your model to ensure your model output is properly normalized.
+    """
+
+    def __init__(
+        self,
+        distance: Distance = "cosine",
```
Comment (owenvallis) [inline tensorflow_similarity/losses/lifted_structure_loss.py:88]: This should be: `distance: Distance | str = "distance"`

You also need to add `from __future__ import annotations` to support the new typing syntax

File: tensorflow_similarity/losses/lifted_structure_loss.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    concatenated_distances = tf.concat([pairwise_distances, reordered_pairwise_distances], axis=1)
+    concatenated_negative_mask = tf.concat([negative_mask, reordered_negative_mask], axis=1)
+    concatenated_negative_mask = tf.cast(concatenated_negative_mask, tf.float32)
+    # Compute (margin - neg_dist) logsum_exp values for each row (equation 4 in the paper)
+    neg_logsumexp = tfsim_losses.utils.logsumexp(margin - concatenated_distances, concatenated_negative_mask)
+
+    # Calculate the loss
+    j_values = neg_logsumexp + positive_dists
+
+    loss = j_values / 2.0
```
Comment (owenvallis) [inline tensorflow_similarity/losses/lifted_structure_loss.py:66]: Seems like mypy isn't sure this is a float tensor. I think we can do the following to force the typing.

`loss: FloatTensor = j_values / 2.0`

**Review by owenvallis — CHANGES_REQUESTED** [review]

**Comment by owenvallis** [comment]
Looks like the tests/losses/test_lifted_structure_loss.py just needs to be formatted and then the PR should be good to go.

**Comment by Lorenzobattistela** [comment]
@owenvallis Sorry took so long, I missed your comment in the notifications. I ran black `black . --verbose --check` in my branch locally, and i got this output:
`Oh no! 💥 💔 💥 14 files would be reformatted, 144 files would be left unchanged.`

But i commited only the change in test_lsl, since it was the file related to this PR. If it is needed to change other things, just tag me and I'll do it. Thanks!

**Comment by owenvallis** [comment]
Looks like there are some isort errors with the import formatting but that shouldn't block us here. I'll format the imports on my side. 

Thanks again for implementing the loss!

**Comment by Lorenzobattistela** [comment]
Appreciate all the help, I'll keep contributing to similarity. Thanks
